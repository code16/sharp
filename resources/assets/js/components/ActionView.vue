<template>
    <div class="SharpActionView">
        <template v-if="showErrorPage">
            <div class="container">
                <h1>Error {{errorPageData.status}}</h1>
                <p>{{errorPageData.message}}</p>
            </div>
        </template>
        <template v-else>
            <slot />
            <notifications position="top right" animation-name="slideRight" style="top:6rem" reverse>
                <template slot="body" slot-scope="{ item, close }">
                    <div class="SharpToastNotification" :class="`SharpToastNotification--${item.type||'info'}`" role="alert" data-test="notification">
                        <div class="SharpToastNotification__details">
                            <h3 class="SharpToastNotification__title mb-2">{{ item.title }}</h3>
                            <p v-if="!!item.text" class="SharpToastNotification__caption" v-html="item.text"></p>
                        </div>
                        <button class="SharpToastNotification__close-button" type="button" @click="close" data-test="close-notification">
                            <svg class="SharpToastNotification__icon" aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd">
                                <path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
                            </svg>
                        </button>
                    </div>
                </template>
            </notifications>
            <template v-for="dialog in dialogs">
                <Modal
                    v-bind="dialog.props"
                    @ok="dialog.okCallback"
                    @hidden="dialog.hiddenCallback"
                    :key="dialog.id"
                >
                    {{ dialog.text }}
                </Modal>
            </template>
        </template>

        <LoadingOverlay :visible="isLoading" />
    </div>
</template>

<script>
    import { api } from "../api";
    import { Modal, LoadingOverlay } from 'sharp-ui';

    export default {
        name:'SharpActionView',
        components: {
            Modal,
            LoadingOverlay,
        },

        provide() {
            return {
                axiosInstance: api
            }
        },

        props: {
            context: {
                type: String,
                required: true
            }
        },

        data() {
            return {
                showErrorPage: false,
                errorPageData: null
            }
        },
        computed: {
            dialogs() {
                return this.$store.state.dialogs;
            },
            isLoading() {
                return this.$store.getters.isLoading;
            },
        },
        created() {
            let { axiosInstance } = this._provided;

            axiosInstance.interceptors.response.use(c=>c, error=>{
                let { response: {status, data}, config: { method } } = error;

                if(method==='get' && status === 404) {
                    this.showErrorPage = true;
                    this.errorPageData = {
                        status, message: data.message
                    }
                }
                return Promise.reject(error);
            });
        }
    }
</script>
