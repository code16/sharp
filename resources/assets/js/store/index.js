
export default {
    state: {
        loading: false,
        dialogs: [],
    },
    mutations: {
        setLoading(state, loading) {
            state.loading = !!loading;
        },
        setDialogs(state, dialogs) {
            state.dialogs = dialogs;
        },
    },
    getters: {
        isLoading(state) {
            return !!state.loading;
        },
    },
    actions: {
        setLoading({ commit }, loading) {
            commit('setLoading', loading);
        },
        setDialogs({ commit }, dialogs) {
            commit('setDialogs', dialogs);
        },
    },
}