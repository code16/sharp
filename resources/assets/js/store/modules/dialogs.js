
export default {
    namespaced: true,

    state: {
        dialogs: [],
    },

    actions: {
        showDialog() {

        },
        hideDialog() {

        },
    },
}