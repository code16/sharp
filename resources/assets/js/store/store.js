import Vuex from "vuex";
import storeModule from './index';

let currentStore = null;

export function store() {
    if(!currentStore) {
        return currentStore = new Vuex.Store(storeModule);
    }
    return currentStore;
}