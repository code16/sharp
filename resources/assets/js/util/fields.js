import Vue from 'vue';
import { logError } from "./log";

const customFieldRE = /^custom-(.+)$/;

export function isCustomField(type) {
    return customFieldRE.test(type);
}

export function resolveCustomField(type) {
    const [_, name] = type.match(customFieldRE) || [];
    const component = name
        ? Vue.options.components[`SharpCustomField_${name}`]
        : null;
    if(!component) {
        logError(`unknown custom field type '${type}', make sure you register it correctly (https://sharp.code16.fr/docs/guide/custom-form-fields.html#register-the-custom-field)`);
    }
    return component;
}