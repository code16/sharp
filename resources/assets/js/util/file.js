import filesize from 'filesize';


function getLocale() {
    return (navigator.language || '').slice(0, 2) || 'en';
}

function getSymbols(locale) {
    if(locale === 'fr') {
        return {
            KB: 'Ko',
            MB: 'Mo',
        }
    }
}

// https://github.com/avoidwork/filesize.js
export function filesizeLabel(bytes) {
    const locale = getLocale();
    const exponent = Math.max(filesize(bytes, { output: 'exponent' }), 1);
    const resolvedBytes = Math.max(bytes, 128);
    const label = filesize(resolvedBytes, {
        standard: 'jedec',
        round: 2,
        exponent,
        locale: true,
        symbols: getSymbols(locale),
    });

    if(bytes < 128) {
        return `< ${label}`;
    }
    return label;
}