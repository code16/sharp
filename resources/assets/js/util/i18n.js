
export function lang(key) {
    return window.i18n[key] || '';
}