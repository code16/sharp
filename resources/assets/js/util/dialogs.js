import { store } from "../store/store";

let modalId = 0;

export function showDialog({ text, okCallback = ()=>{}, okCloseOnly, isError, ...props }) {
    const id = modalId++;
    function hiddenCallback() {
        store().dispatch('setDialogs', store().state.dialogs.filter(dialog => dialog.id !== id));
    }
    store().dispatch('setDialogs', [
        ...store().state.dialogs,
        {
            id,
            props: {
                ...props,
                okOnly: okCloseOnly,
                noCloseOnBackdrop: okCloseOnly,
                noCloseOnEsc: okCloseOnly,
                visible: true,
                isError
            },
            okCallback,
            hiddenCallback,
            text,
        }
    ]);
}

export function showAlert(message, { title, ...props } = {}) {
    return showDialog({
        okCloseOnly: true,
        text: message,
        title,
        ...props
    });
}

export function showConfirm(message, { title, ...props } = {}) {
    return showDialog({
        text: message,
        title,
        ...props
    });
}