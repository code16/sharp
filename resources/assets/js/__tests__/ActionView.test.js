import Vuex from 'vuex';
import Notifications from 'vue-notification';
import { shallowMount, createLocalVue, config } from '@vue/test-utils';
import ActionView from '../components/ActionView.vue';
import { showAlert, showConfirm } from "../util/dialogs";
import store from '../store';

config.stubs['transition-group'] = false;

describe('action-view', ()=>{

    const ModalStub = {
        name:'TestModal',
        template:'<div><slot/></div>'
    };

    const localVue = createLocalVue();
    localVue.use(Notifications);
    localVue.use(Vuex);

    function createWrapper(options={}) {
        return shallowMount(ActionView, {
            ...options,
            propsData: {
                context: 'form',
                ...options.propsData
            },
            stubs: {
                'SharpModal': ModalStub
            },
            store: new Vuex.Store(store),
            localVue
        })
    }

    function notify(wrapper, ...args) {
        wrapper.vm.$notify(...args);
        // remove notification global data-id to preserve unit
        wrapper.findAll('[data-id]').wrappers.forEach(wrapper=>wrapper.element.removeAttribute('data-id'));
    }

    test('can mount ActionView', ()=>{
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('can mount ActionView with notification', ()=>{
        const wrapper = createWrapper();
        notify(wrapper, { title:'TITLE', text:'MESSAGE', type:'TYPE' });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount ActionView with error page', ()=>{
        const wrapper = createWrapper({
            data: ()=>({
                showErrorPage: true,
                errorPageData: {
                    status: 404,
                    message: 'ERROR MESSAGE'
                }
            })
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount with multiple modals', ()=>{
        const wrapper = createWrapper();
        showAlert('Modal 1');
        showConfirm('Modal 2');
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('loading', () => {
        const wrapper = createWrapper();
        wrapper.vm.$store.dispatch('setLoading', true);
        expect(wrapper.html()).toMatchSnapshot();
    });

    xtest('handle main modal events', async ()=>{
        const wrapper = createWrapper();
        const modalOptions = { okCallback:jest.fn() };

        showAlert('Modal 1', modalOptions);

        await wrapper.vm.$nextTick();

        let modal = wrapper.find(ModalStub);

        modal.vm.$emit('ok');
        expect(modalOptions.okCallback).toHaveBeenCalled();
    });
});
