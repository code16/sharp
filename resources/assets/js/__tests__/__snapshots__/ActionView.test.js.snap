// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action-view can mount ActionView 1`] = `
<div class="SharpActionView">
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0" max="Infinity" style="top: 6rem;"></notifications-stub>
	<loadingoverlay-stub></loadingoverlay-stub>
</div>
`;

exports[`action-view can mount ActionView with error page 1`] = `
<div class="SharpActionView">
	<div class="container">
		<h1>Error 404</h1>
		<p>ERROR MESSAGE</p>
	</div>
	<loadingoverlay-stub></loadingoverlay-stub>
</div>
`;

exports[`action-view can mount ActionView with notification 1`] = `
<div class="SharpActionView">
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0" max="Infinity" style="top: 6rem;"></notifications-stub>
	<loadingoverlay-stub></loadingoverlay-stub>
</div>
`;

exports[`action-view can mount with multiple modals 1`] = `
<div class="SharpActionView">
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0" max="Infinity" style="top: 6rem;"></notifications-stub>
	<modal-stub visible="true" okonly="true" nocloseonbackdrop="true" nocloseonesc="true">
		Modal 1
	</modal-stub>
	<modal-stub visible="true">
		Modal 2
	</modal-stub>
	<loadingoverlay-stub></loadingoverlay-stub>
</div>
`;

exports[`action-view loading 1`] = `
<div class="SharpActionView">
	<notifications-stub group="" width="300" reverse="true" position="top right" classes="vue-notification" animationtype="css" animation="[object Object]" animationname="slideRight" speed="300" cooldown="0" duration="3000" delay="0" max="Infinity" style="top: 6rem;"></notifications-stub>
	<modal-stub visible="true" okonly="true" nocloseonbackdrop="true" nocloseonesc="true">
		Modal 1
	</modal-stub>
	<modal-stub visible="true">
		Modal 2
	</modal-stub>
	<loadingoverlay-stub visible="true"></loadingoverlay-stub>
</div>
`;
